src/main/java/com/conveyal/taui/models/ModificationStop.java

```diff
@@ -29,78 +28,88 @@
+        }
+
         // Keep a stack of Stops because as part of auto-generating stops we sometimes need to back one out.
         Stack<ModificationStop> stops = new Stack<>();
-        CoordinateReferenceSystem crs = DefaultGeographicCRS.WGS84;
 
-        if (segments == null || segments.size() == 0) {
-            return new ArrayList<>();
+        // Create an empty array if dwell times are null
+        if (dwellTimes == null) {
+            dwellTimes = new Integer[0];
         }
 
         Segment firstSegment = segments.get(0);
+        int realStopIndex = 0;
 
         if (firstSegment.stopAtStart) {
-            stops.add(new ModificationStop(firstSegment.geometry.getCoordinates()[0],
-                    firstSegment.fromStopId, false, 0));
+            Coordinate c = firstSegment.geometry.getCoordinates()[0];
+            StopSpec stopSpec = new StopSpec(c.x, c.y);
+            if (firstSegment.fromStopId != null) {
+                stopSpec = new StopSpec(firstSegment.fromStopId);
+            }
+
+            Integer specificDwellTime = dwellTimes[realStopIndex++];
+            int dwellTime = specificDwellTime != null ? specificDwellTime : defaultDwellTime;
+
+            stops.add(new ModificationStop(stopSpec, dwellTime, 0, false));
         }
 
         double metersToLastStop = 0; // distance to previously created stop, from start of pattern
         double metersFromPatternStart = 0; // from start of pattern
         int hopTimeSeconds = 0;
+        int previousSegmentSpeedKph = segmentSpeedsKph.length > 0 ? segmentSpeedsKph[0] : DEFAULT_SEGMENT_SPEED_KPH;
+        Coordinate previousSegmentFinalCoordinate = null;
+
+        // Iterate over segments, which are defined by control points and real (i.e. not auto-generated at a requested
+        // spacing) stops.
         for (int i = 0; i < segments.size(); i++) {
             Segment segment = segments.get(i);
-            int segmentSpeedKph = i < segmentSpeedsKph.length ? segmentSpeedsKph[i] : DEFAULT_SEGMENT_SPEED_KPH;
+
+            // A timetable's segment speeds array may get out of sync with the
+            int segmentSpeedKph = i < segmentSpeedsKph.length ? segmentSpeedsKph[i] : previousSegmentSpeedKph;
             double segmentSpeedMps = segmentSpeedKph * METERS_PER_KM / SECONDS_PER_HOUR;
 
+            // Set the previous segment speed to the current one since we just use the `mps` later
+            previousSegmentSpeedKph = segmentSpeedKph;
+
             int stopSpacing = segment.spacing;
             boolean autoCreateStops = stopSpacing > 0;
 
-            // TODO sanity check that each segment's final point matches the first point of the next segment
             Coordinate[] coords = segment.geometry.getCoordinates();
+            if (previousSegmentFinalCoordinate != null && !coords[0].equals2D(previousSegmentFinalCoordinate)) {
+                throw AnalysisServerException.unknown("Segment coordinates do not align.");
```

Comment:
Nice. Is it also worth checking that previous stopAtEnd = current stopAtStart?

Reply:
Possibly, there could be a whole host of sanity checks that may be worth adding in.